homeassistant/helpers/device_utils.py

```diff
@@ -0,0 +1,74 @@
+"""Provides useful helpers for dealing with devices."""
+
+from homeassistant.core import HomeAssistant, callback
+
+from . import device_registry as dr, entity_registry as er
+
+
+@callback
+def async_device_id_to_entity_id(
+    hass: HomeAssistant,
+    entity_id: str,
+) -> str | None:
+    """Resolve the device id to the entity id."""
+
+    ent_reg = er.async_get(hass)
+    if ((entity := ent_reg.async_get(entity_id)) is not None) and (
+        entity.device_id is not None
+    ):
+        return entity.device_id
+    return None
+
+
+@callback
+async def async_device_info_to_link(
+    hass: HomeAssistant,
+    device_id: str | None = None,
+    entity_id: str | None = None,
+) -> dr.DeviceInfo | None:
+    """DeviceInfo with the information to link a device in a config entry in the Link category.
+
+    It can directly receive a device ID or an entity ID to be extracted from the device.
+    """
+
+    dev_reg = dr.async_get(hass)
+
+    if device_id is None and entity_id is not None:
+        device_id = async_device_id_to_entity_id(hass, entity_id=entity_id)
+
+    if (
+        device_id is not None
+        and (device := dev_reg.async_get(device_id=device_id)) is not None
+    ):
+        return dr.DeviceInfo(
+            identifiers=device.identifiers,
+            connections=device.connections,
+        )
+
+    return None
+
+
+@callback
+async def async_remove_stale_device_links_helpers(
+    hass: HomeAssistant,
+    entry_id: str,
+    source_entity_id: str | None = None,
+    device_id: str | None = None,
+) -> None:
+    """Remove obsolete devices from helper config entries that inherit from source entity.
+
+    This can receive an entity ID from which the correct device ID will be extracted or it
+    can directly receive a device ID to be kept in the config entry.
```

Comment:
I dont think we should overload this with features. Keep it simple and just take entity_id. This used once on startup so not a hot path to do optimizations.

If you want you can split it to two functions, one taking entity_id and one taking device_id. The former calling the latter.

Reply:
Still overloaded. Dont have an optional device id. Have one function with a required device id, and one function with a required entity_id. The one with a required entity_id calls the one with a required device id once it knows it.

Reply:
I tried to make the function names as explanatory as possible, but I think it's still not good...
`async_remove_stale_devices_links_keep_entity_device` and `async_remove_stale_devices_links_keep_current_device`

Please, if you have any suggestions that would be great.